internal/gcs/guestconnection.go

```diff
@@ -0,0 +1,129 @@
+package gcs
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/Microsoft/hcsshim/internal/schema1"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Microsoft/go-winio/pkg/guid"
+)
+
+const (
+	protocolVersion = 4
+
+	gcsVsockPort            = 0x40000000
+	firstIoChannelVsockPort = 0x40000001
+
+	nullContainerID = "00000000-0000-0000-0000-000000000000"
+)
+
+type GuestConnection struct {
+	listener  net.Listener
+	brdg      *bridge
+	vmID      guid.GUID
+	mu        sync.Mutex
+	nextPort  uint32
+	notifyChs map[string]chan struct{}
+	caps      schema1.GuestDefinedCapabilities
+	log       *logrus.Entry
+}
+
+// NewGuestConnection creates a new GCS connection for the given VM.
+func NewGuestConnection(vmID guid.GUID, log *logrus.Entry) (*GuestConnection, error) {
+	l, err := winio.ListenHvsock(&winio.HvsockAddr{
+		VMID:      vmID,
+		ServiceID: winio.VsockServiceID(gcsVsockPort),
+	})
+	if err != nil {
+		return nil, err
+	}
+	gc := &GuestConnection{
+		listener:  l,
+		vmID:      vmID,
+		nextPort:  firstIoChannelVsockPort,
+		notifyChs: make(map[string]chan struct{}),
+		log:       log,
+	}
+	return gc, nil
+}
+
+// Capabilities returns the guest's declared capabilities.
+func (gc *GuestConnection) Capabilities() *schema1.GuestDefinedCapabilities {
+	return &gc.caps
+}
+
+// Protocol returns the protocol version that is in use.
+func (gc *GuestConnection) Protocol() uint32 {
+	return protocolVersion
+}
+
+// Connect starts the guest connection.
+func (gc *GuestConnection) Connect(ctx context.Context) error {
+	conn, err := gc.listener.Accept()
+	gc.listener.Close()
+	if err != nil {
+		return err
+	}
+	gc.brdg = newBridge(conn, gc.notify, gc.log)
+	gc.brdg.Start()
+	req := negotiateProtocolRequest{
+		MinimumVersion: protocolVersion,
+		MaximumVersion: protocolVersion,
+	}
+	var resp negotiateProtocolResponse
+	resp.Capabilities.GuestDefinedCapabilities = &gc.caps
+	err = gc.brdg.RPC(ctx, rpcNegotiateProtocol, &req, &resp, true)
+	if err != nil {
```

Comment:
As an entrypoint to handling downlevel Windows GCS's you could handle UnknownMessage here as a special error which would indicate that the guest didnt support the NegotiateProtocol call. We don't have to handle it today but it would be the place we would handle this as a fallback to v3

Reply:
Right, although downlevel GCSs will send a UTF-16 response, so it will be a little more complicated.